### ISSUE TITLE
Simplex Noise functions (2D, 3D, and 4D)

### ISSUE THREAD

**Issue by bdero** [godotengine/godot#1268]
**[Simplex Noise](https://en.wikipedia.org/wiki/Simplex_noise)** (also known as improved Perlin Noise) has many uses in modern game development, as it lies at the heart of procedural generation. Infinite terrain, voxel worlds, clouds, and just about any texture with realistic, natural patterns can be generated using Simplex Noise. It's commonplace in modern game engines.
## Simple examples of simplex noise use:
- **3D simplex noise for 2D animation** _(passing through the third dimension over time)_: http://blog.bdero.me/some-simplex-noise.html
- **2D simplex noise for 3D terrain generation:** http://blog.bdero.me/infinite-simplex-noise-map-3d.html
## Example implementations:

Fortunately, it's also pretty easy to implement.
- **JavaScript implementation:** https://github.com/jwagner/simplex-noise.js/blob/ecfa36aed21d56b298143633b30be554754f3410/simplex-noise.js#L139-L262
- **Python implementation:** https://code.google.com/p/battlestar-tux/source/browse/procedural/simplexnoise.py#243
- **C++ implementation:** https://code.google.com/p/battlestar-tux/source/browse/procedural/simplexnoise.cpp#243

The ideal implementation would involve allowing the user to set the RNG's seed somehow (whether it's set using a separate function in advance or passed as a parameter to the noise functions), and exposing the noise functions in "@GDScript", which might look something like this:

``` python
float simplex2d(x, y)
float simplex3d(x, y, z)
```

**Comment by Spooner** [comment]
You would want 4d noise too, so you can get noise in 3d + time. That said, it makes considerably more sense to use find and use existing glsl code yourself in shaders rather than use in gdscript (where it would be running in c++ underneath). The parallelism of the video card would make the calculations thousands of times faster to be real time.

**Comment by bdero** [comment]
@Spooner Great point with the 4d noise and using shaders whenever possible. Rethinking this, I can't really come up with any real use case where it would be favourable to use simplex noise in CPU time over GPU. It might still be favourable for light one-off things, but even for stuff like voxel engines, generating initial block placement would probably best be done with shaders..

**Comment by zauberparacelsus** [comment]
I'd like to also suggest the addition of Voronoi/Worley noise and Musgrave noise.  Both are useful for procedural generation, especially when used in combination with one another.

**Comment by zauberparacelsus** [comment]
Okay, apparently simplex noise is covered by a patent for implementations using three or more dimensions.  There is, however, an alternative implementation called OpenSimplex that skirts the patent issues, though afaik it is only implemented in Java.

**Comment by zauberparacelsus** [comment]
Found an implementation of opensimplex written in C: https://github.com/smcameron/open-simplex-noise-in-c

**Comment by est31** [comment]
The problem with simplex noise is that its patent protected. One would have to chose an implementation that avoids to be covered by the patent's claims.

**Comment by zauberparacelsus** [comment]
@est31 Which is why I suggested OpenSimplex, which (despite the name) is NOT based off of the patented implementation of Simplex.

**Comment by nunodonato** [comment]
Coming from 2d game dev, I've used noise functions extensively in my games, and never in shaders. So I would have to vote to keep it out of the shader-related stuff :)

**Comment by est31** [comment]
Until its supported natively by godot, there is a 2D/3D simplex implementation in gdscript: https://github.com/OvermindDL1/Godot-Helpers/blob/master/Simplex/Simplex.gd

**Comment by razvanc-r** [comment]
Someone wrote a C version of simplex noise. I think will be of much help with implementing it in the core game engine. https://github.com/smcameron/open-simplex-noise-in-c. If I get time I'll try to make a module for it meanwhile (if no one else beats me to it :d)

**Comment by zauberparacelsus** [comment]
@razvanc-r It is worth noting that it is Open Simplex, not Simplex.  The implementation is different than that of Simplex, and (importantly) is not encumbered by patents.

**Comment by sergicollado** [comment]
It would be nice to have noise functionality in godot :+1:

**Comment by Zylann** [comment]
I'm playing with terrain generation and noise functions would be very useful to speed up GDScript implementations. I tried with shaders but in Godot it feels too limited yet.

I also found this library, FastNoise, which is MIT-licensed: https://github.com/Auburns/FastNoise

I would also point out that not only a noise3d(x,y,z) is good to have, but also noise3d(x,y,z, grid) so we can ask C++ to generate a whole grid of values in various formats (floats, bytes...).
Then I think noise should not be just functions, but objects, so we can set their seed, offsets, storage format, parameters etc

**Comment by Zylann** [comment]
I've just got FastNoise to work in Godot as a module. World generation is blazing fast now :) 

# A glimpse of the usage
var noise = Noise.new()
noise.set_frequency(32)
noise.set_octaves(3)
...
var h = noise.get_noise_3d(x, y, t)

But I read about the patent issues with Simplex noise so I also started to integrate OpenSimplex, following the same GDScript interface.
A notable difference is that OpenSimplex allocates memory for the permutation table *per seed*, which FastNoise doesn't...

**Comment by Zylann** [comment]
I finished an OpenSimplex Godot integration :)
https://github.com/Zylann/godot_opensimplex

**Comment by bdero** [comment]
Thank you for your work @Zylann , I took a look and it looks like a great implementation. The native inclusion of the [fractal noise](https://github.com/Zylann/godot_opensimplex/blob/59b899c19a00c4adc70add16f3cacab3ceba62b7/noise.h#L40-L80) class seems epecially good, since it covers probably the most common use pattern.
If the whole repository is MIT licensed, you may want to consider moving LICENSE.md up to the root of the repo and including the license or a link to the license at the bottom of the readme.

**Comment by Zylann** [comment]
The license of the OpenSimplex C library is public domain.
I added a LICENSE file for the rest of the code :)

**Comment by sergicollado** [comment]
Oh!!! I can't wait to check your wrapper, thanks Zylann!!

**Comment by Zylann** [comment]
Just in case, here is my FastNoise integration I was talking about earlier: https://github.com/Zylann/godot_fastnoise
It has more features and usage is a bit simpler, but it has the Simplex noise patent issue.

**Comment by Zylann** [comment]
Apparently [the patent](https://www.google.com/patents/US6867776) on Simplex noise (referred as Perlin Noise) covers its use for image generation. So using it for terrain generation should be fine, isn't it?

**Comment by est31** [comment]
I am no lawyer, but isn't the terrain displayed on the screen via images (aka "frames")?

So its probably fine if you don't display the generated terrain on an image.

**Comment by razvanc-r** [comment]
Why isn't this bundled with the engine already? :)

**Comment by vnen** [comment]
> Why isn't this bundled with the engine already? :)

Because nobody sent any PR...

**Comment by reduz** [comment]
Lets wait until we can supply binary modules for 3.0, then we bundle this

**Comment by razvanc-r** [comment]
Oh, didn't know there's plan to add binary modules. That would be great cause to recompile it every time... well, that's why I don't use these modules now :). Probably not only me either.

**Comment by hyperglow** [comment]
Hey guys,

Do you think the noise module implementation is usable to generate animated 4d textures? 

Would ne awesome...

**Comment by Zylann** [comment]
My two modules (OpenSimplex and FastNoise for Godot) both provide a 4d simplex noise generator so you can at most animate a 3D texture (the 4th dimension being time).

**Comment by hyperglow** [comment]
The way would be call the function for every pixel and fill a pixel buffer set it as texture? Is gd script fast enough for this?

**Comment by Zylann** [comment]
If you need to do it only once / bake it, GDScript can do the job with a native module (or even a pure script implementation, which will be much slower: https://github.com/OvermindDL1/Godot-Helpers/blob/master/Simplex/Simplex.gd). If you need to do that every frame... you can forget about native modules, and do that with a shader. However I'm not sure if the current Godot shader language has enough features to make it simple :') (also, note that the patent on Simplex noise targets this usage).

**Comment by quantum-entangler** [comment]
All we really need is to be able to do iteration in shaders/shadergraphs and we will be able to implement our own cell/noises in shaders, I mean besides just the basic hash noise which is possible already now.

**Comment by bdero** [comment]
I made a comment much earlier (around the time when I made this issue) about how I couldn't think of non image/hightmap-related use cases for simplex noise, but really there are many. For instance, if you want to make your camera shake a certain way using simplex noise, or you want to simulate the smooth random direction change of an insect as it's moving.

**Comment by razvanc-r** [comment]
This isn't just about shaders, simplex noise (as other types of random functions) can be used for an infinite other purposes, not just visuals. Coherent noise can be done with http://libnoise.sourceforge.net/ for example (which is just another name for something like simplex noise, but doesn't come attached with patent issues etc.)

**Comment by quantum-entangler** [comment]
Yeah I've been eyeing that libnoise myself, and I do agree that noise is useful in more way's than just the one. :)

However my point still stands, as far as shaders are concerned all we need is to be able to do iteration in shaders to generate our own noise, and I'm inclined to say that noise can already be generated in GDScript.
This in no way is meant to say that we couldn't make good use of something like libnoise implementation, though(it would be fantastic, alas no MIT License there - it's LGPL).

**Comment by Zylann** [comment]
Personally, for some game projects I often need more than shaders (in which I actually sample textures, which is faster), and when I do for other purposes it's usually a huge CPU bottleneck when done in GDScript. My requirements are usually the ability to setup seeds and query any point in 1, 2, 3 or 4D infinite seamless space, with preferably as less memory allocations as possible.
I was pretty amazed by FastNoise for this, as it allocates zero memory, is lightweight, provides a few different noises which are not limited in space and generation is very fast.
On the other hand I keep hearing about libnoise but I wonder if it can fulfill the same requirements in game development?

**Comment by quantum-entangler** [comment]
Oh, yeah, fastnoise is a compatible license so a definite +1 to that. :)

**Comment by Zylann** [comment]
If I understood @reduz , OpenSimplexNoise would be a DLScript?
I have no problem with that, if it were about libnoise or complete frameworks...
But how about having the basics in core? It covers lots of use cases by itself. If more complex stuff is needed on more advanced projects, then custom modules or DLScript are better suited for it due to the size it adds.
Or did you mean Godot will be bundled with some modules that will be supported in official export templates rather than being 3rd party?

**Comment by LinuxUserGD** [comment]
https://github.com/Zylann/godot_heightmap_module
@Zylann
Could your module be capable for this implementation? It should be possible to port this code to C++ via GDnative.

**Comment by willnationsdev** [comment]
@Zylann are we thinking your noise module would be integrated for the 3.1 release, or.is this something that would have to be a GDNative submodule for some reason (would prefer a module implementation)?

**Comment by Zylann** [comment]
Given how widespread the usage is, a module would be great. I made an OpenSimplex implementation as a module, feel free to check it out

**Comment by willnationsdev** [comment]
I thought you used FastNoise too though? @Zylann

**Comment by Zylann** [comment]
Would be fine too (might even be better) if you don't mind patents

**Comment by willnationsdev** [comment]
@Zylann How would patents get involved? FastNoise is MIT-licensed, so there shouldn't be any issue in integrating a Noise Reference-type into Godot (I would think).

**Comment by Zylann** [comment]
It's not about the implementation license, but the patent on the method itself. I think it was mentionned earlier in this thread

**Comment by willnationsdev** [comment]
@Zylann Ohhhh, I get it now. So you were wanting to substitute the portions of FastNoise that would infringe on the patent for 2D content with Open Simplex implementations? Or are we just doing Open Simplex all the way?

**Comment by Zylann** [comment]
Put shortly, I prefer FastNoise because it is faster, lighter in memory and has more features. But due to patents I also ported an OpenSimplex module, which is slower, has less features and allocates more memory. I didn't write the libs, and I didn't work on mixing them because I use my free time for other things at the moment :p

**Comment by Zylann** [comment]
So due to that, I proposed OpenSimplex all the way despite my preference, but feem free to tweak the module.

**Comment by Chaosus** [comment]
I suppose this patent is active only in empire of evil(USA) :) ? Because in Russia and Europe patents for program methods does not exist. However if someone decided to push game with patented FastNoise algo to Steam the problems can happens ? However its not easy provably...

**Comment by vnen** [comment]
> However its not easy provably...

This is not the right way to think. And probably false, if they know the game is made in Godot, which has the source available, it's easy to connect the dots.

Patent for software and algorithms does not exist AFAIK, but this case falls into a sort of grey area, since it's described as an "apparatus" which includes a computer and a display. Nevertheless, I'm not a lawyer and don't know all the implications of it, therefore I would be on the safe side. If someone with proper knowledge of the USA law explain why this would be okay, then fine, otherwise I prefer the status quo.

In any case, the patent only covers image generation. The problem is if some user unknowingly uses it to create images in the game, which could cause them trouble and such trouble likely would fall into Godot's lap since they were just using the engine.

**Comment by Chaosus** [comment]
Ah, then having it as separate library is the only option. Pushing slow unpatented version to Godot does not firstly seems like a good idea, but if someone would like to use faster version instead slower for image generation, this fact will be hidden(if not open source) by slow version and hardly proven(I guess Godot binary is not easy for decompile).

**Comment by willnationsdev** [comment]
Well, I'm trying to create a unified implementation that uses FastNoise in most cases, but suddenly switches over to OpenSimplex any time you want to do a 2D simplex noise operation.

**Comment by Zylann** [comment]
@Chaosus technically, OpenSimplex should still be faster than PerlinNoise, only slower than Simplex.

**Comment by Chaosus** [comment]
And maybe performance is not so critical for this, because it called usually several times but before level creation. Not much games uses noise for generate level chunks on fly. And for graphical part of shaders there is already existed assets in Asset library...

**Comment by JosefKuchar** [comment]
Any progress on this so far?

**Comment by aaronfranke** [comment]
If we add 4D noise, it may be worthwhile to add 4D math types first (such as Vector4).

**Comment by OvermindDL1** [comment]
For note, 5D and higher noise has immense uses as well, don't just constrain it to lower dimensions.  It is not hard to create an N-Dimension version of the algorithm (though I did that via C++ templates long long ago so the usages got optimized extremely well, and I think I was using perlin then...).

**Comment by OvermindDL1** [comment]
> For example, a function accepts float arrays, and passing an array of 7 floats = 7D noise.

That's what I did in C++, if it could statically know the size of the array then it generated optimized code, else it fell back to recursion (which depending on what was being done could be about the same speed or a lot slower, depending on how well SSE could be used).

**Comment by JFonS** [comment]
I went ahead and added open-simplex-noise as a Godot module in #21569. Just commenting here in case someone wants to improve on it (maybe add 5+D noise?) since I won't be working on it in the near future. If anyone wants to work on it I'm open to answering questions.

**Comment by Calinou** [comment]
Simplex noise was implemented in #21569, so this issue can now be closed.

Thanks for your contribution @JFonS! :tada:
